Blog of 2.4.0 release (#4677)
apache/pulsar: site2/website/blog/2019-07-05-Apache-Pulsar-2-4-0.md

---
author: Penghui Li
authorURL: https://twitter.com/lipenghui6
title: Apache Pulsar 2.4.0
---

> We are glad to present the new 2.4.0 release of Pulsar. This is the result of a huge 

Reviewer: ```suggestion
We are glad to publish Apache Pulsar 2.4.0 release. This is the result of a huge 
```

effort from the community, with over 460 commits and a long list of new features, 
general improvements and bug fixes.

Check out the <b>[release notes](/release-notes/#2.4.0)</b> for a detailed list of 
the changes, with links to the relevant pull-requests, discussions and documentation.

Regarding new features introduced, I just want to highlight here a tiny subset of them:

<!--truncate-->

### Delayed message delivery

> It's now possible to send delayed message by pulsar producer, and delayed message will 

Reviewer: 1. "... send a delayed message..." or "... send delayed messages..."; please check all similar cases.
2. use "Pulsar" instead of "pulsar". Please check all similar cases.
3. "..., and the delayed message/messages will be..."

available after delay time.

> The Java code for a client using delayed message delivery will look like:

Reviewer: ```suggestion
The Java code for a client using delayed message delivery looks as follows:
```

> The Java code for a client using delayed message delivery will look like:

Reviewer: You can try to use present tense(looks...), instead of future tense(will be...)

```java
producer.newMessage().value("delayed message").deliverAfter(10, TimeUnit.SECONDS).send()
```

> Note:
>
> 1. Messages are only delayed on shared subscriptions. Other subscriptions will deliver immediately.

> > 2. Can't work well with batching messages while messages has different delay time.

Reviewer: Delayed message cannot work well with ...?

Reviewer 2: > Can't work well with batching messages while messages has different delay time.

It does work. It will just send the messages individually.

Reviewer: That's really great news.

### Go functions

> Before 2.4.0, pulsar support use Java/Python to write pulsar functions. Now, it's 

Reviewer: ```suggestion
Before 2.4.0 release, Java and Python are supported to write Pulsar Functions. Now, it's 
```

> possible to use golang to write pulsar functions, the following is an example of 

Reviewer: "Pulsar Functions"
Please check all similar cases.

a Pulsar Function written in golang.

```go
import (
    "fmt"
    "context"

    "github.com/apache/pulsar/pulsar-function-go/pf"
)

func HandleRequest(ctx context.Context, in []byte) error {
    fmt.Println(string(in) + "!")
    return nil
}

func main() {
    pf.Start(HandleRequest)
}
```

### Key_shared subscription

> A new subscribe type `Key_shared`. By `Key_shared` one partition could have several 

Reviewer: It's better to re-write this paragraph.

consumers to parallel consume messages, while all messages with the same key will be 
dispatched to only one consumer ordered. 
Here is [architecture](http://pulsar.apache.org/docs/en/concepts-messaging/#key_shared) 
for Key_Shared.

The following is an example to use `Key_shared` subscription:

```java
client.newConsumer()
        .topic("topic")
        .subscriptionType(SubscriptionType.Key_Shared)
        .subscriptionName("sub-1")
        .subscribe();
```

### Schema versioning

> Before 2.4.0, avro schema is using one schema for both its writer schema and reader schema. 

Reviewer: ```suggestion
Before 2.4.0 release, avro schema uses one schema for both writer schema and reader schema. 
```

> Multiple schema version is support now.

Reviewer: supported

> Use multiple schema, producer can send messages with different schema version and consumer 

Reviewer: You can use "With multiple schema" or "Using multiple schema", either is right.

can read messages with different schema.

> In 2.4.0, added `FORWARD_TRANSITIVE`, `BACKWARD_TRANSITIVE` and `FULL_TRANSITIVE` compatibility 

Reviewer: ```suggestion
In 2.4.0 release, `FORWARD_TRANSITIVE`, `BACKWARD_TRANSITIVE` and `FULL_TRANSITIVE` compatibility strategies are added to check ...
```

Reviewer: Could you check all similar cases in the following context?

strategy to  check the compatibility with all existing schema version.

### Replicated subscription

In 2.4.0, added a mechanism to keep subscription state in-sync, within a sub-second timeframe, 
in the context of a topic that is being asynchronously replicated across multiple geographical 
regions. Here is [architecture](https://github.com/apache/pulsar/wiki/PIP-33%3A-Replicated-subscriptions) 
for replicated subscription.

The following is an example to use replicated subscription:

```java
Consumer<String> consumer = client.newConsumer(Schema.STRING)
            .topic("my-topic")
            .subscriptionName("my-subscription")
            .replicateSubscriptionState(true)
            .subscribe();
```

### New IO connectors

> A new batch of connectors was added, including Flume, Redis sink, Solr sink, RabbitMQ sink, 

Reviewer: "was added" . ---> "is added"

InfluxDB sink. Here is list of builtin [connectors](http://pulsar.apache.org/docs/en/io-connectors/) 

> that pulsar already supported now.

Reviewer 3: ```suggestion
that Pulsar already supports now.
```

### Security

In 2.4.0, added support Kerberos in pulsar broker and client. 
Following the [document](http://pulsar.apache.org/docs/en/security-kerberos/) to enable Kerberos authentication.

## Conclusion

> Please [download](/download) Pulsar 2.4.0 and report feedback, issues or any comment into our mailing lists,

Reviewer 3: ```suggestion
If you want to download Pulsar 2.4.0, click [here](https://pulsar.apache.org/en/download/).
You can send any questions or suggestions to our mailing lists, contribute to Pulsar on [GitHub](https://github.com/apache/pulsar) or join the Apache Pulsar community on [Slack](https://apache-pulsar.herokuapp.com/).
```

### Go Functions

Before 2.4.0 release, Java and Python are supported to write Pulsar Functions. Now, you can 
use Go to write Pulsar Functions, the following is an example of 
a Pulsar Function written in Go.

```go
import (
    "fmt"
    "context"

    "github.com/apache/pulsar/pulsar-function-go/pf"
)

func HandleRequest(ctx context.Context, in []byte) error {
    fmt.Println(string(in) + "!")
    return nil
}

func main() {
    pf.Start(HandleRequest)
}
```

### Key_shared subscription

A new subscription mode `Key_shared` is introduced in 2.4.0. In `Key_shared` subscription mode, 
one partition could have several consumers to consume messages in parallelism and ensure messages 
with the same key are distributed to a consumer in order. 
Here is [architecture](http://pulsar.apache.org/docs/en/concepts-messaging/#key_shared) 
for Key_Shared.

The following is an example to use `Key_shared` subscription:

```java
client.newConsumer()
        .topic("topic")
        .subscriptionType(SubscriptionType.Key_Shared)
        .subscriptionName("sub-1")
        .subscribe();
```

### Schema versioning

> Before 2.4.0 release, avro schema uses one schema for both its writer schema and reader schema. 

Reviewer 3: ```suggestion
Before 2.4.0 release, Avro schema used one schema for both writer schema and reader schema. 
```

Multiple schemas version is supported now.

> With multiple schemas, producer can send messages with different schema version and consumer 

Reviewer 3: ```suggestion
With multiple schemas, a producer can send messages with different schema versions and a consumer 
```

> can read messages with different schema.

Reviewer 3: ```suggestion
can read messages with different schemas.
```

In 2.4.0 release, `FORWARD_TRANSITIVE`, `BACKWARD_TRANSITIVE` and `FULL_TRANSITIVE` compatibility 
strategies are added to check the compatibility with all existing schema version.

### Replicated subscription

In 2.4.0 release, a mechanism is added to keep subscription state in sync, within a sub-second timeframe, 
in the context of a topic that is being asynchronously replicated across multiple geographical 
regions. Here is [architecture](https://github.com/apache/pulsar/wiki/PIP-33%3A-Replicated-subscriptions) 
for replicated subscription.

The following is an example to use replicated subscription:

```java
Consumer<String> consumer = client.newConsumer(Schema.STRING)
            .topic("my-topic")
            .subscriptionName("my-subscription")
            .replicateSubscriptionState(true)
            .subscribe();
```

### New IO connectors

A new batch of connectors is added, including Flume, Redis sink, Solr sink, RabbitMQ sink. 
The following lists builtin [connectors](http://pulsar.apache.org/docs/en/io-connectors/) 

> that Pulsar supported.

Reviewer: supports

### Security

> In 2.4.0 release, added support Kerberos in Apache Pulsar broker and client. 

Reviewer: ```suggestion
In 2.4.0 release, Kerberos is supported in Apache Pulsar broker and client. 
```

> Following the [document](http://pulsar.apache.org/docs/en/security-kerberos/) to enable Kerberos authentication.

Reviewer: ```suggestion
To enable Kerberos authentication, refer to the [document](http://pulsar.apache.org/docs/en/security-kerberos/).
```

Also added role based Pulsar Function authentication and authorization.

## Conclusion

> If you want download Pulsar 2.4.0, click [here](/download). You can send any questions or suggestions 

Reviewer 3: ```suggestion
If you want to download Pulsar 2.4.0, click [here](/download). You can send any questions or suggestions 
```

> If you want download pulsar 2.4.0, click [here](/download). You can send any questions or suggestions 

Reviewer: Pulsar

> Note:
>
> 1. Messages are only delayed on shared subscriptions, other subscriptions will deliver immediately.

> > 2. Delayed messages are send individually even if you enable message batching on producer.

Reviewer: sent
